File: Examenes/ExamenPSP/2DAM_PSP-main/9 - Caja fuerte asimétrica/CajaFuerte/cliente/src/main/java/ui/screens/principal/PrincipalController.java
package ui.screens.principal;


import common.CachedCredentials;
import jakarta.enterprise.inject.Instance;
import jakarta.inject.Inject;
import javafx.fxml.FXML;
import javafx.fxml.FXMLLoader;
import javafx.scene.control.Alert;
import javafx.scene.control.ButtonType;
import javafx.scene.control.MenuBar;
import javafx.scene.control.MenuItem;
import javafx.scene.layout.BorderPane;
import javafx.scene.layout.Pane;
import javafx.stage.Stage;
import javafx.stage.WindowEvent;
import lombok.extern.log4j.Log4j2;
import org.utils.domain.modelo.Folder;
import org.utils.domain.modelo.User;
import ui.screens.common.BaseScreenController;
import ui.screens.common.ScreenConstants;
import ui.screens.common.Screens;

import java.io.IOException;
import java.util.Optional;

@Log4j2
public class PrincipalController extends BaseScreenController {

    @FXML
    private MenuItem menuLogout;

    // objeto especial para DI
    Instance<Object> instance;

    @FXML
    private MenuBar menuPrincipal;

    private Stage primaryStage;


    @FXML
    public BorderPane root;


    private final Alert alert;

    private User userLoggedIn;

    private Folder folderLoggedIn;

    private final CachedCredentials ca;



    @Inject
    public PrincipalController(Instance<Object> instance, CachedCredentials ca) {
        this.instance = instance;
        alert = new Alert(Alert.AlertType.NONE);
        this.ca = ca;
    }


    public void showAlertError(String mensaje) {
        alert.setAlertType(Alert.AlertType.ERROR);
        alert.setContentText(mensaje);
        alert.setResizable(true);
        alert.showAndWait();
    }

    public void showAlertInformation(String message) {
        Alert infoAlert = new Alert(Alert.AlertType.INFORMATION);
        infoAlert.setContentText(message);
        infoAlert.showAndWait();
    }

    private Pane cargarPantalla(String ruta) {
        Pane panePantalla = null;
        try {

            FXMLLoader fxmlLoader = new FXMLLoader();
            fxmlLoader.setControllerFactory(controller -> instance.select(controller).get());
            panePantalla = fxmlLoader.load(getClass().getResourceAsStream(ruta));
            BaseScreenController pantallaController = fxmlLoader.getController();
            pantallaController.setPrincipalController(this);
            pantallaController.principalCargado();


        } catch (IOException e) {
            log.error(e.getMessage(), e);
        }
        cambioPantalla(panePantalla);
        return panePantalla;
    }


    private void cambioPantalla(Pane pantallaNueva) {
        root.setCenter(pantallaNueva);
    }


    public void initialize() {
        goToLogin();
    }


    public void cargarPantalla(Screens pantalla) {
        cambioPantalla(cargarPantalla(pantalla.getRuta()));
    }

    private void closeWindowEvent(WindowEvent event) {
        Alert alerta = new Alert(Alert.AlertType.INFORMATION);
        alerta.getButtonTypes().remove(ButtonType.OK);
        alerta.getButtonTypes().add(ButtonType.CANCEL);
        alerta.getButtonTypes().add(ButtonType.YES);
        alerta.setTitle(ScreenConstants.QUIT_APPLICATION);
        alerta.setContentText(ScreenConstants.CLOSE_WITHOUT_SAVING);
        alerta.initOwner(primaryStage.getOwner());
        Optional<ButtonType> res = alerta.showAndWait();


        res.ifPresent(buttonType -> {
            if (buttonType == ButtonType.CANCEL) {
                event.consume();
            }
        });


    }

    public void exit() {
        primaryStage.fireEvent(new WindowEvent(primaryStage, WindowEvent.WINDOW_CLOSE_REQUEST));
    }

    public void setStage(Stage stage) {
        primaryStage = stage;
        primaryStage.addEventFilter(WindowEvent.WINDOW_CLOSE_REQUEST, this::closeWindowEvent);
    }



    public void onLoginHecho(User user) {
        userLoggedIn = user;
        menuPrincipal.setVisible(true);
        menuLogout.setVisible(true);
        cargarPantalla(Screens.MAIN_FOLDERS);
    }

    public void backToLogin() {
        goToLogin();
    }

    private void goToLogin() {
        logout();
        menuPrincipal.setVisible(true);
        cargarPantalla(Screens.LOGIN.getRuta());
    }

    private void logout() {
        userLoggedIn = null;
        folderLoggedIn = null;
        ca.invalidate();
    }

    public void goToSignUp() {
        logout();
        cargarPantalla(Screens.REGISTER);
    }

    public User getUserLoggedIn() {
        return userLoggedIn;
    }

    public void goToFolderMessages(Folder folder) {
        folderLoggedIn = folder;
        cargarPantalla(Screens.MAIN_MESSAGES.getRuta());
    }

    public Folder getFolderSelected() {
        return folderLoggedIn;
    }
}

File: Examenes/ExamenPSP/2DAM_PSP-main/11 - Spring GraphQL/graphqlCliente/src/main/kotlin/ui/controllers/PrincipalController.kt
package ui.controllers

import credentials.CachedCredentials
import domain.modelo.Persona
import javafx.fxml.FXML
import javafx.fxml.FXMLLoader
import javafx.scene.control.*
import javafx.scene.layout.BorderPane
import javafx.scene.layout.Pane
import javafx.stage.Stage
import javafx.stage.WindowEvent
import lombok.extern.log4j.Log4j2
import ui.common.BaseScreenController
import ui.common.Screens
import java.io.IOException
import kotlin.system.exitProcess

@Log4j2
class PrincipalController {


    @FXML lateinit var menuAdmin: Menu
    @FXML lateinit var menuUser: Menu
    @FXML lateinit var menuLogout: MenuItem

    @FXML
    lateinit var menuPrincipal: MenuBar

    private var primaryStage: Stage? = null


    @FXML
    var root: BorderPane? = null


    private var alert: Alert = Alert(Alert.AlertType.NONE)

    var userLoggedIn: Persona? = null


    private var ca: CachedCredentials = CachedCredentials









    fun showAlertError(mensaje: String?) {
        alert.alertType = Alert.AlertType.ERROR
        alert.contentText = mensaje
        alert.isResizable = true
        alert.showAndWait()
    }

    fun showAlertInformation(message: String?) {
        val infoAlert = Alert(Alert.AlertType.INFORMATION)
        infoAlert.contentText = message
        infoAlert.showAndWait()
    }

    private fun cargarPantalla(ruta: String): Pane? {
        var panePantalla: Pane? = null
        try {
            val fxmlLoader = FXMLLoader()
            panePantalla = fxmlLoader.load(javaClass.getResourceAsStream(ruta))
            val pantallaController: BaseScreenController = fxmlLoader.getController()
            pantallaController.setPrincipalController(this)
            pantallaController.principalCargado()
        } catch (e: IOException) {
            showAlertError("Error al cargar la pantalla: $ruta ($e)")
        }
        cambioPantalla(panePantalla)
        return panePantalla
    }


    private fun cambioPantalla(pantallaNueva: Pane?) {
        root!!.center = pantallaNueva
    }


    fun initialize() {
        goToLogin()
    }


    private fun cargarPantalla(pantalla: Screens) {
        cambioPantalla(cargarPantalla(pantalla.ruta))
    }

    private fun closeWindowEvent(event: WindowEvent) {
        val alerta = Alert(Alert.AlertType.INFORMATION)
        alerta.buttonTypes.remove(ButtonType.OK)
        alerta.buttonTypes.add(ButtonType.CANCEL)
        alerta.buttonTypes.add(ButtonType.YES)
        alerta.title = "Quit application?"
        alerta.contentText = "Close without saving?"
        alerta.initOwner(primaryStage!!.owner)
        val res = alerta.showAndWait()
        res.ifPresent { buttonType: ButtonType ->
            if (buttonType == ButtonType.CANCEL) {
                event.consume()
            } else{
                //Parar ejecución sin inyección de dependencias
                primaryStage!!.close()
                exitProcess(0)
            }
        }
    }

    fun exit() {
        primaryStage?.fireEvent(WindowEvent(primaryStage, WindowEvent.WINDOW_CLOSE_REQUEST))
    }

    fun setStage(stage: Stage?) {
        primaryStage = stage
        primaryStage!!.addEventFilter(
            WindowEvent.WINDOW_CLOSE_REQUEST
        ) { event: WindowEvent ->
            closeWindowEvent(
                event
            )
        }
    }


    fun onLoginHecho(user: Persona) {
        userLoggedIn = user
        menuPrincipal.isVisible = true
        menuLogout.isVisible = true
        menuUser.isVisible = true
        menuAdmin.isVisible = ca.role.equals("ROLE_ADMIN")
        cargarPantalla(Screens.LIST_MASCOTAS)
    }

    fun backToLogin() {
        goToLogin()
    }

    private fun goToLogin() {
        logout()
        menuPrincipal.isVisible = true
        cargarPantalla(Screens.LOGIN)
    }

    private fun logout() {
        userLoggedIn = null
        ca.invalidate()
        menuUser.isVisible = false
        menuAdmin.isVisible = false
        menuLogout.isVisible = false
    }

    fun goToSignUp() {
        logout()
        cargarPantalla(Screens.REGISTER)
    }

    fun showLoading(show: Boolean = true) {
        //Set cursor to wait
        if (show) {
            root!!.cursor = javafx.scene.Cursor.WAIT
        } else {
            root!!.cursor = javafx.scene.Cursor.DEFAULT
        }
    }

    fun goToMyProfile() {
        cargarPantalla(Screens.MY_PROFILE)
    }

    fun goToMyMascotas() {
        cargarPantalla(Screens.LIST_MASCOTAS)
    }

    fun goToListPersons() {
        cargarPantalla(Screens.LIST_PERSONAS)
    }

}